src/common/libeventlog/eventlog.c

```diff
@@ -0,0 +1,331 @@
+ *
+ * SPDX-License-Identifier: LGPL-3.0
+\************************************************************/
+
+#if HAVE_CONFIG_H
+#include "config.h"
+#endif
+#include <flux/core.h>
+#include <jansson.h>
+#include <czmq.h>
+#include <argz.h>
+
+#include "eventlog.h"
+
+static json_t *eventlog_entry_decode_wrapper (const char *event,
+                                              bool trailing_newline);
+
+int eventlog_entry_parse (json_t *entry,
+                          double *timestamp,
+                          const char **name,
+                          json_t **context)
+{
+    double t;
+    const char *n;
+    json_t *c;
+
+    if (!entry) {
+        errno = EINVAL;
+        return -1;
+    }
+
+    if (json_unpack (entry, "{ s:F s:s }",
+                     "timestamp", &t,
+                     "name", &n) < 0) {
+        errno = EINVAL;
+        return -1;
+    }
+
+    if (timestamp)
+        (*timestamp) = t;
+    if (name)
+        (*name) = n;
+
+    if (!json_unpack (entry, "{ s:o }", "context", &c)) {
+        if (context)
+            (*context) = c;
+    }
+    else {
+        if (context)
+            (*context) = NULL;
+    }
+
+    return 0;
+}
+
+static int eventlog_entry_append (json_t *a, const char *event)
+{
+    json_t *o = NULL;
+
+    if (!(o = eventlog_entry_decode_wrapper (event, false)))
+        return -1;
+
+    if (json_array_append_new (a, o) < 0) {
+        json_decref (o);
+        errno = ENOMEM;
+        return -1;
+    }
+
+    return 0;
+}
+
+json_t *eventlog_decode (const char *s)
+{
+    char *copy = NULL;
+    char *input;
+    json_t *a = NULL;
+    int save_errno;
+    char *ptr;
+
+    if (!s) {
+        errno = EINVAL;
+        return NULL;
+    }
+
+    if (!(copy = strdup (s)))
+        goto enomem;
+    input = copy;
+
+    /* gotta have atleast 1 newline, if not empty string */
+    if (strlen (s) && !(ptr = strchr (input, '\n'))) {
+        errno = EINVAL;
+        goto error;
+    }
+
+    if (!(a = json_array ()))
+        goto enomem;
+
+    while ((ptr = strchr (input, '\n'))) {
+        (*ptr++) = '\0';
+
+        if (eventlog_entry_append (a, input) < 0)
+            goto error;
+
+        input = ptr;
+    }
+
+    free (copy);
+    return a;
+
+ enomem:
+    errno = ENOMEM;
+ error:
+    save_errno = errno;
+    free (copy);
+    json_decref (a);
+    errno = save_errno;
+    return NULL;
+}
+
+bool eventlog_entry_validate (json_t *entry)
+{
+    json_t *name;
+    json_t *timestamp;
+    json_t *context;
+
+    if (!json_is_object (entry)
+        || !(name = json_object_get (entry, "name"))
+        || !json_is_string (name)
+        || !(timestamp = json_object_get (entry, "timestamp"))
+        || !json_is_number (timestamp))
+        return false;
+
+    if ((context = json_object_get (entry, "context"))) {
+        if (!json_is_object (context))
+            return false;
+    }
+
+    return true;
+}
+
+static json_t *eventlog_entry_decode_wrapper (const char *entry,
```

Comment:
Different name?  It's not a wrapper for event_decode(), it's the other way around right?